README.rst
==========
mboxfilter
==========

`mboxfilter <http://pamoller.com/mboxfilter.html>`_ is a Python class for filtering and sorting emails in mailboxes (mbox-format) by values of `header fields <http://tools.ietf.org/html/rfc5322#section-3.6>`_. You may find it useful for archiving and reporting. Basic filtering an sorting can be done by the command line tool mboxfilter, which comes with mboxfilter and acts as wrapper around the class. All emails send from Peter to Frank are grepped from a bunch of mailboxes easyly:

::

    $ mboxfilter --filter_from "peter@(?:work.com|home.org)" --filter_to frank@work.com peter2012.mbox peter2013.mbox

All filters accept regular expressions. In the example above they are matched to the values of the header fields From and To. An email passes a filter if every filter item is passed. The output is printed to stdout by default. To sort all emails sent from Peter by reciver and year is done by:

::

    $ mboxfilter --filter_from "peter@home.org" --sort_to --sort_date %Y peter2012.mbox peter2013.mbox
    $ ls
    frank@home.org.2012.mbox
    frank@home.org.2013.mbox
    rosie@home.org.2012.mbox ...

Sorting may produces different result set. Each result set is written to a own file. Note: An email wirtten to Frank and Rosie is added to two result sets - as shown above. Sorting by Date accepts a `format string <http://docs.python.org/2/library/datetime.html#strftime-and-strptime-behavior>`_ as argument. mboxfilter can also be used for archiving emails. All emails of Peter are archived by:

::

    $ mboxfilter --archive peter@home.org.mbox peter@work.com.mbox
    $ ls 
    2012.mbox
    2013.mbox
    index.sqlite3

Important header values of the emails are bundeled in a row and added to the index database index.sqlite3. The index revokes emails from being added twice. But the SQLite 3 database is also meant for querying by SQL.

More complex tasks can be done by using the Python class. For finding all emails sent from Peter to Rosie (To, Cc or Bcc) is done by:

::

    #!/usr/bin/env python
    from mboxfilter import Filter
    import sys
    cached = []
    for header in ["To", "Cc", "Bcc"]:
      f = Filter(caching=True, filters=[("From", sys.argv[0]), (header, sys.argv[1])])
      f.filter(sys.argv[2])
      cached = cached + f.cache
    res = Filter(selectors=[("Date", "%Y")])
    res.filter(cached) 

=====
Class
=====

::

    class mboxfilter.Filter(output ::= "./", archive ::= False, indexing ::= False, filters ::= [], selectors ::= [], caching ::= False, separator ::= ".")

The Filter class can be used to instantiate own filters. The parameter filters takes a list of tuples, e.g.: [("From", "peter@home.org"), ("To", "rosie@home.org")]. The first item of the tuple references the name of a header field. It's value is matched against the regular expression within the second item. (The matches are kept, see below) An email passes a  filter if every filter item is passed (matched). That are all emails from Peter to Rosie in the preceding example.


The tuples, which are listed in the parameter selectors, sort the emails into different result sets. The selector tuples are structured like the filter tuples, but without a regular expressions. The second value is ignored, except for the Date filed. The Date field expects a `format string <http://docs.python.org/2/library/datetime.html#strftime-and-strptime-behavior>`_ to extract a approbative date string from the Date field. A list of selectors may look like: [("To", None), ("Date", "%Y")]. The selectors constructs for every email a set of sort keys. A sort key is a string, which consists of a sequence of key parts separated by the value of parameter separator. A key part is made from the header's value by the following rules. a)  From, To, Cc, Bcc: a list of email-addresses, b) Date: a user formatted representation of the date and c)  any other: the first 12 chars and the md5 value of the header value. Possible sort keys for the example above are: rosie@home.org.2012 and rosie@home.org.2013.

Note: Header fields like From, To, Cc or Bcc may create more than one sort key per email, e.g. emails sent to two recivers with filters=[] and selector=[("To", None)]: The list of recivers forms two result sets: rosie@home.org and frank@work.com.

Note: If a header field was used as filter before, e.g. filters=[('To', 'rosie@home.org')] and selectors=[('To', None)], the matches will be used to from the sort key instead of the list of recivers: rosie@home.org only.

Without any selector the filtered emails are printed to stdout. Otherwise the sort keys are used as file names. The parameter output can be used to write the files into an existing directory. The parameter indexing=True creates the result set database in the file index.sqlite3. The Parameter archive=True is a shorthand for the parameter combination: indexing=True and selectors=[("Date", "%Y")]. It is useful for adding mails once in an archive. The parameter caching=True redirects the output to the class member cache. This setting disables  indexing and ignores any selector.

=======
Members
=======

::

    filtered ::= 0

Number of emails that has been filtered. 

::

    passed ::= "."

Number of emails, that has passed the filters. 

::

    cache ::= []

List of passed emails, if parameter caching was set.

=======
Methods
=======

::

     filter_mbox(obj)

filter_mbox apply all filters and selectors to every email in the referenced mailbox. The Parameter obj can be either an instance of the mailbox.mbox class or a path referring to a mailbox within the mbox format.

::

     filter_mail(mail)

filter_mail apply all filters and selectors to the parameter mail, an instance of the mailbox.mboxMessage class.

===
Cmd
===

::

    mboxfilter [--help] [--version] [--nostat] [--dir output] [--unique] [--archive] [--filter_from regexp] [--filter_to regexp] [--filter_date regexp] [--filter header,regexp] [--sort_from] [--sort_to] [--sort_date format] [--sort header,regexp] mbox ...

For simple tasks the wrapper mboxfilter can be used as shown in the beginning. To use any header field for filtering or for sorting the options filter or sort are used. Note: the name of the header field and the regexp (Regular Expression) are separated by a ",". The option nostat suppresses the printing of a statistic to stderr.

CHANGES
Changelog for mboxfilter
------------------------

0.1.0
-----

Initial commint


0.1.3
-----

handle only real headers

test/test_mboxfilter.py
# -*- coding: UTF-8 -*-
import email
import email.mime.text
import email.header
import mailbox
import mboxfilter
import os
import re
import shutil
import sys
import unittest

SUBJ    = 'L\xf6rem ips\xfcm dolor sit amet'
BODY    = 'Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy  eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam  nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.'
DATE_1  = 'Friday, 21 June 2013 12:15:00 +0200'
DATE_2  = 'Sunday, 23 June 2013 07:15:22 +0200'
DATE_3  = 'Tuesday, 25 June 2013 18:41:49 +0300'
DATE_4  = 'Wednesday, 26 June 2013 18:42:00 +0000'
DATE_5  = 'Friday, 28 June 2013 03:13:59 +0200'
MSGID_1 = 'mssgid-1'
MSGID_2 = 'mssgid-2'
MSGID_3 = 'mssgid-3'
MSGID_4 = 'mssgid-4'
MSGID_5 = 'mssgid-5'
DIR     = 'unit-tests'
PERS_1  = 'Frank Heinrich W\xf6ller'
MAIL_1  = "fheini@woeller.de"
ADDR_1  = '"' + PERS_1 + '" <' + MAIL_1 + '>'
PERS_2  = 'Friederich Claus Sch\xfc\xdfler'
MAIL_2  = 'fclaus@schuessler.de'
ADDR_2  = '"' + PERS_2 + '" <' + MAIL_2 + '>'
MAIL_3  = 'anna-lisa@meissner-jakobi.de'
PERS_3  = 'Anna Lisa Mei\xdfner-Jakobi'
ADDR_3  = '"' + PERS_3 + '" <' + MAIL_3 + '>'
PERS_4  = 'Jos\xe9 Carlos García Fuentes'
MAIL_4  = 'jose@garcia.es'
ADDR_4  = '"' + PERS_4 + '" <' + MAIL_4 + '>'
PERS_5  = ''
MAIL_5  = 'spammer@ugly.org'
ADDR_5  = MAIL_5
MBOX    = DIR + '/' + 'mbox'
MBOX_1  = MBOX + '_1'
MBOX_2  = MBOX + '_2'
MBOX_3  = MBOX + '_3'

ERROR_DATE_NOT_FOUND = 'header "Date" not found in mail'
ERROR_FROM_NOT_FOUND = 'header "From" not found in mail'
ERROR_CANT_ADD_TWICE = "can't add mail twice to result index" 
class MboxFilterOutput(mboxfilter.Filter):
  """ Test classes redirect output to stack unit."""
  def __init__(self, output=None, archive=False, indexing=False, filters=[], selectors=[], caching=False, separator="."):
    mboxfilter.Filter.__init__(self, output, archive, indexing, filters, selectors, caching, separator)
    self.unit = []
    self.unit_failures = []

  def resultset_add(self, mail):
    """ append mail to a resultset by sort keys """
    if self.sort_keys_generate(mail) > 0:
      for key in self.sort_keys:
        self.unit.append((mail, key))
    else:
      self.unit.append((mail, None))

  def error(self, msg, mail):
    self.unit_failures.append((mail, msg))

class MboxFilterOutputIntegrity(MboxFilterOutput):
  def error(self, exc, mail):
    self.unit.append((mail, exc[0])) 

class TestMboxFilter(unittest.TestCase):
  def __init__(self, nop):
    unittest.TestCase.__init__(self, nop)
    shutil.rmtree(DIR, True)
    os.mkdir(DIR)
    self.mbox1()

  def test_header_value_formatted(self):
    fil = mboxfilter.Filter()	   
    self.assertEqual(MAIL_1, mboxfilter.header_value_formatted('From', None, ADDR_1))
    self.assertEqual(MAIL_5, mboxfilter.header_value_formatted('From', None, ADDR_5))

  #def test_header_decode(self):
  #  fil = mboxfilter.Filter()                                                               
  #  self.assertEqual(fil.header_strip(ADDR_2, '"') + ', '+ MAIL_3.decode('iso-8859-1'), fil.header_decode('''=?utf-8?Q?Friederich_Claus_Sch=C3=BC=C3=9Fler?= <fclaus@schuessler.de>,
  #anna-lisa@meissner-jakobi.de'''))

  def test_archive(self):
    fil = MboxFilterOutput(output="unit-tests", archive=True)
    fil.filter_mbox(MBOX_1)
    m = mailbox.mbox(MBOX_1)	  
    self.check_results([(m[5], ERROR_CANT_ADD_TWICE), (m[6], ERROR_DATE_NOT_FOUND)], fil.unit_failures)


  #def test_no_filter(self): todo fix 
  #  fil = MboxFilterOutput()
  #  fil.filter_mbox(MBOX_1)
  #  m = mailbox.mbox(MBOX_1)
  #  print len(m), fil.unit_failures[0][1]
  #  self.check_results_defaults(m, fil.unit)
  #  self.check_results_defaults([], fil.unit_failures)

  def test_one_filter(self):
    fil = MboxFilterOutput(filters=[("From", PERS_1)])
    fil.filter_mbox(MBOX_1)
    m = mailbox.mbox(MBOX_1)
    self.check_results_defaults([m[0], m[1]], fil.unit)
    self.check_results([(m[6], ERROR_FROM_NOT_FOUND)], fil.unit_failures)

  def test_two_filter(self):
    fil = MboxFilterOutput(filters=[("From", PERS_1), ("From", PERS_2)])
    fil.filter_mbox(MBOX_1)
    m = mailbox.mbox(MBOX_1)
    self.check_results_defaults([m[1]], fil.unit)
    self.check_results([(m[6], ERROR_FROM_NOT_FOUND)], fil.unit_failures)

  def test_many_filter(self):
    fil = MboxFilterOutput(filters=[("From", PERS_1), ("To", PERS_3), ("Message-ID", MSGID_2)])
    fil.filter_mbox(MBOX_1)
    m = mailbox.mbox(MBOX_1)
    self.check_results_defaults([m[1]], fil.unit)
    self.check_results([(m[6], ERROR_FROM_NOT_FOUND)], fil.unit_failures)

  def test_date_sort(self):
    fil = MboxFilterOutput(selectors=[("Date", "%Y")])
    fil.filter_mbox(MBOX_1)
    m = mailbox.mbox(MBOX_1)
    self.check_results_defaults([m[0], m[1], m[2], m[3], m[4], m[5]], fil.unit, str(2013))
    self.check_results([(m[6], ERROR_DATE_NOT_FOUND)], fil.unit_failures)

  def test_from_sort(self):
    fil = MboxFilterOutput(selectors=[("From", None)])
    fil.filter_mbox(MBOX_1)
    m = mailbox.mbox(MBOX_1)
    self.check_results([(m[0], MAIL_1), (m[1], MAIL_1), (m[1], MAIL_2), (m[2], MAIL_3), (m[3], MAIL_2), (m[4], MAIL_4), (m[5], MAIL_4)], fil.unit)
    self.check_results([(m[6], ERROR_FROM_NOT_FOUND)], fil.unit_failures)
  
  def test_from_to_sort(self):
    fil = MboxFilterOutput(selectors=[("From", None), ("To", None)])
    fil.filter_mbox(MBOX_1)
    m = mailbox.mbox(MBOX_1)
    self.check_results([(m[0], MAIL_1+"."+MAIL_2), (m[1], MAIL_1+"."+MAIL_3), (m[1], MAIL_2+"."+MAIL_3), (m[2], MAIL_3+"."+MAIL_2), (m[2], MAIL_3+"."+MAIL_1), (m[3], MAIL_2+"."+MAIL_1), (m[4], MAIL_4+"."+MAIL_1), (m[4], MAIL_4+"."+MAIL_2), (m[5], MAIL_4+"."+MAIL_1), (m[5], MAIL_4+"."+MAIL_2)], fil.unit)
    self.check_results([(m[6], ERROR_FROM_NOT_FOUND)], fil.unit_failures)#

  def test_from_filter_from_sort(self):
    fil = MboxFilterOutput(selectors=[("From", None)], filters=[("From", PERS_1)])
    fil.filter_mbox(MBOX_1)
    m = mailbox.mbox(MBOX_1)
    self.check_results([(m[0], MAIL_1), (m[1], MAIL_1)], fil.unit)
    self.check_results([(m[6], ERROR_FROM_NOT_FOUND)], fil.unit_failures)

  def test_from_to_sort(self):
    fil = MboxFilterOutput(filters=[("From", PERS_1), ("To", PERS_3)], selectors=[("From", None), ("Date", "%a")])
    fil.filter_mbox(MBOX_1)
    m = mailbox.mbox(MBOX_1)
    self.check_results([(m[1], MAIL_1+".Sun")], fil.unit)
    self.check_results([(m[6], ERROR_FROM_NOT_FOUND)], fil.unit_failures)

  #def test_caching(self):
  #  fil_1 = mboxfilter.Filter(caching=True, filters=[("From", PERS_1)])
  ##  fil_2 = mboxfilter.Filter(caching=True, filters=[("To", PERS_1)])    
  #  fil_3 = MboxFilterOutput(output="unit-tests", selectors=[("Date", "%Y")])
  #  fil_1.filter_mbox(MBOX_1)
  #  fil_2.filter_mbox(MBOX_1)
  #  fil_3.filter_mbox(fil_1.cache + fil_2.cache)
  #  m = mailbox.mbox(MBOX_1)
  #  self.check_results([(m[0], "2013"), (m[1], "2013"), (m[2], "2013"), (m[3], "2013"), (m[4], "2013"), (m[5], "2013")], fil_3.unit)

  """  
  def test_subject_sort(self):
    fil = MboxFilterOutput(selectors=[("Subject", None)])
    fil.filter_mbox(MBOX_1)
    m = mailbox.mbox[MBOX_1]
    self.check_results([(m[0], MAIL_1), (m[1], MAIL_1), (m[1], MAIL_2), (m[2], MAIL_3), (m[3], MAIL_2), (m[4], MAIL_4), (m[5], MAIL_4)], fil.unit)
  """
  def check_results_defaults(self, shld, res, default=None):
    self.check_results([(value, default) for value in shld], res)

  def check_results(self, shld, res):
    """ Compare a list of results to """
    self.assertEqual(len(shld), len(res), "leaking results ")
    for i in range(0, len(res)):
      if i < len(shld):
        self.assertTrue(str(shld[i][0]) == str(res[i][0]))
        if (shld[i][1] is not None):
	      # todo check results
          self.assertTrue(self.decode_python(shld[i][1]) == res[i][1])

  def decode_python(self, strg):
    """ Decodes a string if neccessary """
    try: # python 2.6, 2.7
      return strg.decode('iso-8859-1')
    except: # pyhton 3
      return strg

  def mbox1(self):
    """ Create a mbox for testing """
    m = mailbox.mbox(MBOX_1)
    m.add(self.mail(From=[ADDR_1], To=[ADDR_2], Message_ID=MSGID_1, Date=DATE_1))
    m.add(self.mail(From=[ADDR_1, ADDR_2], To=[ADDR_3], Message_ID=MSGID_2, Date=DATE_2))
    m.add(self.mail(From=[ADDR_3], To=[ADDR_2, ADDR_1], Message_ID=MSGID_3, Date=DATE_3))
    m.add(self.mail(From=[ADDR_2], To=[ADDR_1], Message_ID=MSGID_4, Date=DATE_4))
    m.add(self.mail(From=[ADDR_4], To=[ADDR_1, ADDR_2], Message_ID=MSGID_5, Date=DATE_5))
    m.add(self.mail(From=[ADDR_4], To=[ADDR_1, ADDR_2], Message_ID=MSGID_5, Date=DATE_5))
    m.add(self.mail())
    m.flush()

  def mail(self, From=None, To=None, Date=None, Subject=None, Body=BODY, Message_ID=None):
    """ Create a mail """
    mail = email.mime.text.MIMEText(Body)
    if From:
      mail['From'] =  self.header_encode(From)
    if To:
      mail['To']   =  self.header_encode(To)
    if Subject:
      mail['Subject'] = self.header_encode(Subject)
    if Date:
      mail['Date'] = Date
    if Message_ID:
      mail['Message-ID'] = '<' + Message_ID + '>'
    return mail	  

  def header_encode(self, headers):
    """ Encode a header value in ISO-8859-1 """
    return email.header.Header(", ".join(headers), 'iso-8859-1')
	          
if __name__ == '__main__':
  unittest.main()

mboxfilter.py
"""
Filter and sort mails from mboxes for archiving and reporting
"""
import dateutil
import dateutil.parser
import getopt
import hashlib
import email
import email.generator
import mailbox
import os
import re
import sqlite3
import sys
import traceback # todo remove

__version__ = "0.1.4"

# Default encoding of emails:
EMAIL_ENCO = "ISO-8859-15"
# Deafult date format
DATE_FORMAT = "%Y"

class DirectoryNotExisting(Exception):
  def __init__(self, name):
    self.name = name
  def __str__(self):
    return "can't find direcotry: "+self.name

class HeaderMissed(Exception):
  def __init__(self, name):
    self.name = name
  def __str__(self):
    return "header \""+self.name+"\" not found in mail"

class EmailMissed(Exception):
  def __str__(self):
    return "can't find an email address"

class Filter:
  # Output files to directory:
  output = "./"
  # Filname of index database:
  resultset_index = "index.sqlite3"
  # Number of passed mails:
  passed = 0
  # Number of filtered mails:
  filtered = 0
  # Failed while processing:
  failed = 0
  # Do sorting by default:
  sort_date_default = DATE_FORMAT
  # Do indexing by default:
  indexing = False
  # Keep filter matches in List:
  filter_matches = []
  # Keep sort keys in list:
  sort_keys = []  
  # Keep passed mails, when caching:
  passed_mails = []
  # Keep failed mails, when caching:
  failed_mails = []
  # Path of Resultset of failed mails:
  failures_path = None 
   
  def __init__(self, output=None, archive=False, indexing=False, filters=[], selectors=[], caching=False, separator=".", failures=None):
    """ Initialize a Filter object.
      output
        files are outputed into directory
        
      archive
        use default sorting and indexeing
        
      indexing
        import mail headers into result index database named index.sqlite3 table mails. duplets are revoked.
        
      filters
         List of tuples, e.g. ("From", regexp), ("To", regexp) or ("Date", format)
         
      selectors
         List of tuples, e.g. ("From", None), ("To", None), ("Date", format). mails will be output to files.

      caching
        cache resultset. Disables output and indexing
      
      separator
        separates key parts
        
      nostat
        supress statistics
        
      failures path
        add mails, than can not be handeled (e.g. missing headers) to resultset path
    """
    self.output = output or "."
    if not os.path.isdir(self.output):
      raise DirectoryNotExisting(self.output)
    self.archive = archive
    self.filters = filters
    self.selectors = selectors
    if self.archive and len(self.selectors) == 0:
      self.selectors.append(("Date", self.sort_date_default))
    if self.archive or indexing:
      self.indexing = True
      self.index_init()
    self.caching = caching
    # Backward compatibility:
    self.cache = []#self.passed_mails
    self.sort_key_separator = separator
    self.failures_path = failures

  def filter_mbox(self, obj):
    """ Filter a mailbox instance. """
    if isinstance(obj, str):
      if os.path.isfile(obj):
        obj = mailbox.mbox(obj)
    for mail in obj:
      self.filter_mail(mail)
    if isinstance(obj, mailbox.mbox):
      obj.close()

  def filter_mail(self, mail):
    """ Filter a single mail. """
    try:
      self.filtered += 1
      if (self.filter_mail_pass(mail)):
        if self.indexing: # and not caching # disables indexing
          self.index_add(mail)
        self.resultset_add(mail)
        self.passed += 1
    except sqlite3.IntegrityError as excp:
	  #todo add to own resultset!
      self.error("can't add mail twice to result index", mail)
    except HeaderMissed as excp:
      #todo add to own result set
      self.error(str(excp), mail)
    except EmailMissed as excp:
      self.error(str(excp), mail)
    except: # todo unkown error needed?
      self.failed += 1
      sys.stderr.write("unkonwn error - can't handle mail\n")

  def error(self, msg, mail):
    """ Output an error. """
    self.failed += 1
    sys.stderr.write("error: "+msg+" "+self.signature(mail)+"\n")
    self.error_switch(mail)

  #todo test
  def signature(self, mail):
    """ Output a readable signature for a mail. """
    sig = ""
    if "Message-ID" in mail.keys():
      sig += "Message-ID: "+header_value_formatted("Message-ID", None, mail["Message-ID"])+" "
    if "Date" in mail.keys():
      sig += "Date: "+header_value_formatted("Date", DATE_FORMAT, mail["Date"]) 
    return sig
      
  #todo test
  def error_switch(self, mail):
    """ Add mail to resultset of errors. """
    if self.caching:
      self.failed_mails.append(mail)
    elif self.failures_path:
      self.resultset_output(open(os.path.normpath(self.failures_path), "a"), mail)
    #else: todo think about
    #  self.resultset_output(sys.stderr, mail)
  
  def filter_mail_pass(self, mail):
    """ Apply all filter rules. """
    self.filter_matches = {}
    boolean = True
    for header, regexp in self.filters:
       inner_boolean = False
       for header_value in header_values(header, mail):
		  # True if any header part is true:  
         inner_boolean |= self.filter_item_pass(header, regexp, header_value)
        # True if all filter items are true:
       boolean &= inner_boolean
    return boolean
           
  def filter_item_pass(self, header, regexp, strg):
    """ Apply a filter rule. """
    if type(regexp) is not None and len(regexp) > 0:
      if re.search(regexp, strg):
        self.filter_matches_add(header, strg)
        return True
      return False
    return True

  def filter_matches_add(self, key, value):
    """ Keep match of filter items."""
    if key in self.filter_matches.keys():
      self.filter_matches[key].append(value)
    else:
      self.filter_matches[key]=[value]

  def resultset_add(self, mail):
    """ Append mail to a result set. """
    if self.sort_keys_generate(mail) > 0:
      for key in self.sort_keys:
        handle = open(os.path.normpath(self.output + "/" + key + ".mbox"), "a")
        self.resultset_switch(handle, mail)
    else:
      self.resultset_switch(sys.stdout, mail)

  def resultset_switch(self, handle, mail):
    """ Cache mail or write mail to a result set. """
    if self.caching:
      self.cache.append(mail)
    else:
      self.resultset_output(handle, mail)

  def resultset_output(self, handle, mail):
    """ Write mail to a result set. """
    genr = email.generator.Generator(handle, True, 0)
    genr.flatten(mail, True)	  
 
  def sort_keys_generate(self, mail):
    """ Generate the sort keys (result sets) for mail. """
    self.sort_keys = []
    for key, form in self.selectors:
      # Sort by filter matches only (1:1):
      if key in self.filter_matches.keys():
        self.sort_keys_add(key, form, self.filter_matches[key])
      # Sort by all header parts (1:N):
      else:
        self.sort_keys_add(key, form, header_values(key, mail))
    return len(self.sort_keys)

  def sort_keys_add(self, key, form, values):
    """ Append key part to sort keys (generates n*m sort keys for n and m key parts)."""
    new_keys = []
    for value in values:
      key_value = header_value_formatted(key, form, value) 
      # revoke empty key parts:
      if len(key_value) > 0:
        # No keys existing:
        if len(self.sort_keys) == 0:
          new_keys.append(key_value)
        else:
          for sort_key in self.sort_keys:
            new_keys.append(sort_key + self.sort_key_separator + key_value)
      else:
       sys.stderr.write("[NOTE] empty key part\n")
    # Reset to extended set of sort keys:
    self.sort_keys = new_keys
    
  def index_init(self):
    """ Initialize the result index database. """
    self.db = sqlite3.connect(self.index_path())
    self.db.execute('CREATE TABLE IF NOT EXISTS Mails ("MD5-Value" TEXT PRIMARY KEY, "Message-ID" TEXT, "From" TEXT NOT NULL, "To" TEXT NOT NULL, "Cc" TEXT, "Bcc", TEXT, Date TEXT NOT NULL, "In-Reply-To" TEXT, Subject TEXT)');

  def index_add(self, mail):
    """ Add mail header to result index. """ 
    self.db.execute('INSERT INTO Mails ("MD5-Value", "Message-ID", "From", "To", "Cc", "Bcc", Date, "In-Reply-To", Subject) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)', (self.index_md5_value(mail), header_strip(mail['Message-ID']), header_decode(mail['From']), header_decode(mail['To']), header_decode(mail['CC']), header_decode(mail['BCC']), header_decode(mail['Date']), header_strip(mail['In-Reply-To']), header_decode(mail['Subject'])))
    self.db.commit()

  def index_md5_value(self, mail):
    """ Determine a MD5 value for mail. """
    return md5_value(str(mail["Message-ID"] or "") + str(mail["Date"] or "") + str(mail["From"] or "") + str(mail["To"] or ""))
    
  def index_path(self):
    """ Determine the to the result index. """
    return os.path.normpath(self.output + "/" + self.resultset_index)

def md5_value(strg):
  """ Returns the md5 value in hex-fromat of strg """
  md5 = hashlib.md5()
  md5.update(strg.encode("UTF-8"))
  return md5.hexdigest()
	
def header_decode(strg):
  """ Decode header field. """
  if type(strg) is str:
    decoded = []
    for field in email.header.decode_header(strg):
      decoded.append(python_decode(field[0], field[1] or EMAIL_ENCO))
    return "".join(decoded)
  return ""

def header_email(strg):
  """ Filter first email from string """
  match = re.search('([a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,4})', strg)
  if match == None:
    raise EmailMissed()
  return match.group(1)
    
def header_strip(strg, strip='[<>"]'):
  """ Remove encloseing brackets"""
  return re.sub(strip, '', strg or '')

def header_values(header, mail):
  """ Split header into a list of items """
  if header not in mail.keys():
    raise HeaderMissed(header)
  value = header_decode(mail[header])
  if header == "Date":
    return [value]
  elif header in ["From", "Cc", "Bc", "To", "Sender", "Reply-to"]:
    return value.split(",")
  return [value]

def header_value_formatted(key, form, value=""):
  """ Return formatted header values. """
  if key in ["From", "Cc", "Bcc", "To", "Sender", "Reply-to"]:
    return header_email(value)
  elif key == "Date":
    return dateutil.parser.parse(value).strftime(form or DATE_FORMAT)
  # shorten sort key to 20 chars:
  if key == "Message-ID":
    return header_strip(value)
  if len(value or "") > 0: #todo think about
    return (value or "")[:12] + "." + md5_value(value or "") 
  return value

def python_decode(strg, enc):
  """ Decode strings for python < 3. """
  try:
    return strg.decode(enc)
  except:
    return strg

def cli_protocol(val):   
  m = re.search("^([^,;$ ]+)(?:,(.*)|()$)", val)      
  if m:
    return (m.group(1), m.group(2) or "")
  raise "[Error] protocol error"

def cli_info():
  sys.stderr.write("mboxfilter v"+__version__+"\n")

def cli_usage():
  sys.stderr.write("""Usage:
  mboxfilter [--help] [--version] [--dir output] [--unique] [--archive]
  [--filter_from regexp] [--filter_to regexp] [--filter_date regexp]
  [--filter header,regexp] [--sort_from] [--sort_to] [--sort_date format]
  [--sort header,regexp] [--errors path] [--nostat]
  mbox ...\n""")
	
def cli():
  """ Invoke mboxfilter from cmd."""
  try:
    opts, args = getopt.getopt(sys.argv[1:], None, ["dir=", "unique", "archive", "sort_from", "filter_from=", "sort_date=", "filter_date=", "filter_to=", "sort_to", "filter=", "sort=", "help", "nostat", "version", "failures="])
    output = None
    selectors = []
    archive = False
    filters = []
    unique = False
    nostat = False
    failures = None
    for opt, val in opts:
      val = python_decode(val, sys.stdin.encoding)
      if opt == "--dir":
        output = val
      elif opt == "--archive":
       archive = True
      elif opt == "--unique":
       unique = True
      elif opt == "--help":
        cli_usage()
        sys.exit(0)
      elif opt == "--version":
        cli_info()
        sys.exit(0)
      elif opt == "--filter_from":
        filters.append(("From", val))
      elif opt == "--filter_to":
        filters.append(("To", val))
      elif opt == "--filter_date":
        filters.append(("Date", val))
      elif opt == "--filter":
        filters.append(cli_protocol(val))
      elif opt == "--sort_from":
        selectors.append(("From", None))
      elif opt == "--sort_to":
        selectors.append(("To", None))
      elif opt == "--sort_date":
        selectors.append(("Date", val))
      elif opt == "--sort":
        selectors.append(cli_protocol(val))
      elif opt == "nostat":
        nostat = True
      elif opt == "failures":
        failures = val
    filt = Filter(output=output, archive=archive, indexing=unique, filters=filters, selectors=selectors, failures=failures)
    for mbox in args:
      filt.filter_mbox(mbox)
    if not nostat:
      sys.stderr.write(str(filt.filtered) + " mails filtered, " + str(filt.passed) + " passed "+str(filt.failed)+" in " + str(filt.output) + "\n")
  except getopt.GetoptError as excp:
    sys.stderr.write(str(excp)+"\n\n")
    cli_usage()
    sys.exit(1)
  except DirectoryNotExisting as excp:
    sys.stderr.write(str(excp))
    sys.exit(1)
  except SystemExit:
    pass
  except:
    traceback.print_tb(sys.exc_info()[2])
    sys.stderr.write(str(sys.exc_info()[1]));
    sys.exit(1)